### PR TITLE
Allow to configure smtp and sendmail in gitlab.yml

### PR DESCRIPTION
**Pull request by jirutka**
Since almost all GitLab settings are configured in gitlab.yml, it’s very inconvenient to configure SMTP server in `config/initializers/smtp_settings.rb`.

I’ve named it explicitly `outgoing_emails`, because I assume that you may someday implement an ability to respond to emails from GitLab and then you’ll need to distinguish these settings.

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://jirutka-email-settings.c68bfdc4cac4c32c41e1a922e73062ea.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://jirutka-email-settings.c68bfdc4cac4c32c41e1a922e73062ea.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by dserodio** [comment]
I haven't tried the actual PR, but :+1: on the feature itself

**Comment by jvanbaarsen** [comment]
@jirutka Can you please make sure the items raised by Hound are fixed?

**Comment by jirutka** [comment]
@jvanbaarsen Can you please read the source code to see that I’m just consistent with the current code-style?!

File: config/initializers/1_settings.rb [hunk 1]
```diff
@@ -125,6 +123,20 @@ def verify_constant(modul, current, default)
 Settings.gitlab['repository_downloads_path'] = File.absolute_path(Settings.gitlab['repository_downloads_path'] || 'tmp/repositories', Rails.root)
 
 #
+# Outgoing emails
+#
+Settings['outgoing_emails'] ||= Settingslogic.new({})
+Settings.outgoing_emails['enabled']           ||= Settings.gitlab['email_enabled'] unless Settings.gitlab['email_enabled'].nil?  # for backward compatibility
```
Comment (jvanbaarsen) [inline config/initializers/1_settings.rb]: This is not consistend. Please fix.

File: config/initializers/1_settings.rb [hunk 2]
```diff
@@ -125,6 +123,20 @@ def verify_constant(modul, current, default)
...
+Settings.outgoing_emails['enabled']           ||= true if Settings.outgoing_emails['enabled'].nil?
```
Comment (jvanbaarsen) [inline config/initializers/1_settings.rb]: This is not consistend. Please fix

File: config/initializers/disable_email_interceptor.rb [hunk 3]
```diff
@@ -1,2 +1,2 @@
 # Interceptor in lib/disable_email_interceptor.rb
-ActionMailer::Base.register_interceptor(DisableEmailInterceptor) unless Gitlab.config.gitlab.email_enabled
+ActionMailer::Base.register_interceptor(DisableEmailInterceptor) unless Gitlab.config.outgoing_emails.enabled
```
Comment (jvanbaarsen) [inline config/initializers/disable_email_interceptor.rb]: Line need to be shortened.

File: config/initializers/1_settings.rb [hunk 4]
```diff
@@ -125,6 +123,20 @@ def verify_constant(modul, current, default)
...
+Settings.outgoing_emails['from']              ||= Settings.gitlab['email_from'] || "gitlab@#{Settings.gitlab.host}"
+Settings.outgoing_emails['delivery_method']   ||= :sendmail
```
Comment (jvanbaarsen) [inline config/initializers/1_settings.rb]: Please fix this hound warning.

File: config/initializers/1_settings.rb [hunk 5]
```diff
@@ -125,6 +123,20 @@ def verify_constant(modul, current, default)
...
+Settings.outgoing_emails['sendmail_settings'] ||= {}
+Settings.outgoing_emails['smtp_settings']     ||= {}
```
Comment (jvanbaarsen) [inline config/initializers/1_settings.rb]: Please fix.

File: config/initializers/1_settings.rb (hunk 2 above)
Comment (jirutka) [inline config/initializers/1_settings.rb]: No, it is consistent, see [below](https://github.com/jirutka/gitlabhq/blob/email-settings/config/initializers/1_settings.rb#L142-L145).

Comment (jvanbaarsen) [inline config/initializers/1_settings.rb]: Yeah but we decided on a styleguide. I'm asking you to follow it. If not i'm forced to close this PR. I'm sorry!

**Comment by jirutka** [comment]
@jvanbaarsen Rebased to master and all Hound items fixed.

**Comment by jvanbaarsen** [comment]
@maxlazio Do we also need to change stuff in Omnibus for this?

**Comment by maxlazio** [comment]
@jvanbaarsen Yes. 

@jirutka Example case: I am a user who used a manual installation and created the `config/initializers/smtp_settings.rb`. I filled it with the required settings. I upgrade to GitLab 7.11 where this PR was merged. What happens then, will my emails be sent?

**Comment by jirutka** [comment]
@maxlazio No, it will not be sent, you must put these settings to `gitlab.yml`. It’s less evil than trying to make it backward compatible with custom `smtp_settings.rb`.

**Comment by maxlazio** [comment]
If someones GitLab is no longer sending emails due to an upgrade I don't find it lesser evil. I do like the change. @randx What do you think?

**Comment by randx** [comment]
I also like the change. @jirutka can you please make it mergeable?

**Comment by jirutka** [comment]
@randx Done. :shipit:

**Comment by randx** [comment]
@maxlazio I merge this. Please create an issue for omnibus to bring all this settings

**Comment by randx** [comment]
@jirutka thank you

**Comment by jirutka** [comment]
@randx :smile_cat:

**Comment by maxlazio** [comment]
@randx @jirutka Ok, I've created an issue. However this got merged too fast https://gitlab.com/gitlab-org/omnibus-gitlab/issues/555, CHANGELOG item was not included and this is a breaking change.

**Comment by jacobvosmaer** [comment]
I agree with @maxlazio this change has a large impact on people who use an installation from source with SMTP.

@jirutka could you add upgrade instructions to doc/update/7.10-to-7.11.md ?

Besides the CHANGELOG, this change also deserves a mention in the [7.11 release blog post](https://gitlab.com/gitlab-com/www-gitlab-com/merge_requests/670) (cc @JobV).

**Comment by randx** [comment]
Ok I missed part of conversation about backward compatibility. I dont want people suffer during update. We need to make old initialiser work

**Comment by randx** [comment]
@jirutka I will revert this PR because I dont want to break people SMTP settings. But I like the idea behind PR. If you could submit this PR with compatibility fix - it will make us happy :smile:

**Comment by jirutka** [comment]
@randx I preserved backward compatibility for old settings in `gitlab.yml` (see [here](https://github.com/gitlabhq/gitlabhq/blob/b4be7aed7711e51becb3be29ed82d6753cc73cc4/config/initializers/1_settings.rb#L138-L142)). I don’t think that it’s worth it to preserve backward compatibility even for `smtp_settings.rb`.

I’ve sent this PR **three months ago** and already rebased and fixed it two times. I don’t want to rebase it once again…

**Comment by maxlazio** [comment]
@jirutka There are hundreds of installations which need to update this, impact is large. We need to do this properly.

**Comment by jirutka** [comment]
I’ll write update instructions. It’s just about moving settings from `smtp_settings.rb` to `gitlab.yml`, so what’s the problem? Beside that, there is already instruction to check differences in gitlab.yml.example in every update manual:

> There are new configuration options available for [`gitlab.yml`](config/gitlab.yml.example). View them with the command below and apply them to your current `gitlab.yml`.

**Comment by jirutka** [comment]
Since the patch has been reverted, this PR should be reopened IMHO…

**Comment by jvanbaarsen** [comment]
@jirutka I can't open a merged PR, you need to submit a new PR, would be great if you ca link it to this one.

**Comment by jacobvosmaer** [comment]
(the merge commit was reverted so Git will not let us re-merge the same commits (SHAs))

**Comment by jvanbaarsen** [comment]
@jacobvosmaer What do you advice in this situation?

**Comment by jacobvosmaer** [comment]
@jvanbaarsen I think @jirutka found a workaround.

Generally speaking, when your branch has become unmergeable because of a reverted merge, all you can do is somehow create new commits in a new branch, e.g. using rebase.

**Pull request merged**